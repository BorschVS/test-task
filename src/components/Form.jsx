import { Button, TextField } from "@mui/material";
import { useFormik } from "formik";
import { FormStyled } from "styled/Form.styled";
import * as Yup from 'yup';

export const Form = () => {

    const formik = useFormik({
      initialValues: {
        name: '',
        email: ''
      },
      validationSchema: Yup.object({
        name: Yup.string()
          .max(15, 'Must be 15 characters or less')
          .required('Required'),
        email: Yup.string()
          .email('Invalid email address')
          .required('Required')
      }),
      onSubmit: values => {
        alert(JSON.stringify(values, null, 2));
      }
    });

    return (
      <FormStyled onSubmit={formik.handleSubmit}>
        <TextField
          fullWidth
          margin="normal"
          id="name"
          name="name"
          label="Name"
          value={formik.values.name}
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          error={formik.touched.name && Boolean(formik.errors.name)}
          helperText={formik.touched.name && formik.errors.name}
        />
        <TextField
          fullWidth
          margin="normal"
          id="email"
          name="email"
          label="Email Address"
          value={formik.values.email}
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          error={formik.touched.email && Boolean(formik.errors.email)}
          helperText={formik.touched.email && formik.errors.email}
        />
        <Button color="primary" variant="contained" fullWidth type="submit">
          Submit
        </Button>
      </FormStyled>
    );
  };