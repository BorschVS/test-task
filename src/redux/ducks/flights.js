import { takeEvery, put, call } from '@redux-saga/core/effects';

import { FAST_VALUE, CHEAP_VALUE } from 'constants';
import { getSearchId, getFlightsData } from 'api';

const GET_FLIGHTS = 'GET_FLIGHTS';
const SET_FLIGHTS = 'SET_FLIGHTS';
const SET_FLIGHTS_LOADING = 'SET_FLIGHTS_LOADING';
const SET_FLIGHTS_ERROR = 'SET_FLIGHTS_ERROR';
const SET_STOPS_FILTER = 'SET_STOPS_FILTER';
const SET_ALL_STOPS = 'SET_ALL_STOPS';
const SET_STOPS_FILTER_STATUS = 'SET_STOPS_FILTER_STATUS';
const SET_CATEGORY_FILTER = 'SET_CATEGORY_FILTER';

const initialState = {
    availableFlights: [],
    filteredFlights: [],
    stopsFilter: [],
    stopsFilterStatus: false,
    flightCategoryFilter: CHEAP_VALUE,
    loading: false,
    error: null
};

export const flights = (state = initialState, { type, payload }) => {
    switch (type) {
        case SET_FLIGHTS:
            return {
                ...state,
                availableFlights: Array.isArray(payload) ? payload.slice()
                    .sort((prev, next) => prev.price - next.price) : [],
            };
        case SET_FLIGHTS_LOADING:
            return {
                ...state,
                loading: payload
            };
        case SET_FLIGHTS_ERROR:
            return {
                ...state,
                error: payload,
                loading: false
            };
        case SET_CATEGORY_FILTER:
            if (payload === CHEAP_VALUE) {
                return {
                    ...state,
                    flightCategoryFilter: payload,
                    filteredFlights: state.filteredFlights
                        .slice()
                        .sort((prev, next) => prev.price - next.price),
                    availableFlights: state.availableFlights
                        .slice()
                        .sort((prev, next) => prev.price - next.price),
                };
            }
            if (payload === FAST_VALUE) {
                return {
                    ...state,
                    flightCategoryFilter: payload,
                    filteredFlights: state.filteredFlights
                        .slice()
                        .sort(
                            (prev, next) =>
                                prev.segments[0].duration - next.segments[0].duration
                        ),
                    availableFlights: state.availableFlights
                        .slice()
                        .sort(
                            (prev, next) =>
                                prev.segments[0].duration - next.segments[0].duration
                        ),
                };
            }
            return state;

        case SET_STOPS_FILTER:
            return {
                ...state,
                stopsFilter: Array.isArray(payload) ? payload : [],
                filteredFlights: state.availableFlights.filter((flight) =>
                    flight.segments.some(
                        (segment) =>
                            Array.isArray(payload) && payload.includes(segment.stops.length)
                    )
                ),
                stopsFilterStatus: false,
            };
        case SET_ALL_STOPS:
            return {
                ...state,
                stopsFilter: [0, 1, 2, 3],
                filteredFlights: state.availableFlights,
                stopsFilterStatus: true,
            };
        case SET_STOPS_FILTER_STATUS:
            return {
                ...state,
                stopsFilterStatus: Boolean(state.stopsFilter.length),
            };
        default:
            return state;
    }
};

export const getFlights = () => ({
    type: GET_FLIGHTS,
});

export const setFlights = (payload) => ({
    type: SET_FLIGHTS,
    payload,
});

export const setFlightsLoading = (payload) => {
    return {
        type: SET_FLIGHTS_LOADING,
        payload
    };
}

export const setFlightsError = (payload) => {
    return {
        type: SET_FLIGHTS_ERROR,
        payload
    };
}

export const setStopsFilter = (payload) => ({
    type: SET_STOPS_FILTER,
    payload,
});

export const setStopsFilterStatus = () => ({
    type: SET_STOPS_FILTER_STATUS,
});

export const setAllStops = () => ({
    type: SET_ALL_STOPS,
});

export const setCategoryFilter = (payload) => ({
    type: SET_CATEGORY_FILTER,
    payload,
});

// saga

// Worker
export function* handleAvailableFlights() {
    try {
        yield put(setFlightsLoading(true));

        const searchId = yield call(getSearchId);
        const flights = yield call(getFlightsData, searchId);

        yield put(setFlights(flights));
        yield put(setFlightsLoading(false));

    } catch (error) {
        yield put(setFlightsError(error.message));
        yield put(setFlightsLoading(false));
    }
}

// Watcher
export function* watchFlightsSaga() {
    yield takeEvery(GET_FLIGHTS, handleAvailableFlights);
}