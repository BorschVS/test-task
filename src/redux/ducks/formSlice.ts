import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { FormState } from 'types/interfaces';

export const SET_FORM_DATA = 'SET_FORM_DATA';

const initialState: FormState = {
  name: '',
  phone: '',
  email: '',
};

const formSlice = createSlice({
  name: 'form',
  initialState,
  reducers: {
    setFormData(state, action: PayloadAction<FormState>) {
      state.name = action.payload.name;
      state.phone = action.payload.phone;
      state.email = action.payload.phone;
    },
  },
});

export const { setFormData } = formSlice.actions;

export default formSlice.reducer;
