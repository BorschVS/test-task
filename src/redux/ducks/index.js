export { default as flights } from './flights';
export { default as form } from './form';
export { default as modal } from './modal';
export { default as images } from './images';