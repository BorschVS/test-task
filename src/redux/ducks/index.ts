export { flightsReducer } from './flights';
export { formReducer } from './form';
export { modalReducer } from './modal';
export { imagesReducer } from './images';
