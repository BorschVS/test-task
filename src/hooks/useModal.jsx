import { useDispatch, useSelector } from 'react-redux';
import { setIsShowing } from '../redux/ducks/modal';

export const useModal = () => {
  const dispatch = useDispatch();
  const isShowing = useSelector((state) => state.modal.isShowing);

  const toggleModal = () => {
    dispatch(setIsShowing(!isShowing));
  };

  return {
    isShowing,
    toggleModal,
  };
};
