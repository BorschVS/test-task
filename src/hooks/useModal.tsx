import { useDispatch, useSelector } from 'react-redux';
import { setIsShowing } from '../redux/ducks/modal';
import { ReduxState } from 'interfaces';

export const useModal = () => {
  const dispatch = useDispatch();
  const isShowing = useSelector((state: ReduxState) => state.modal.isShowing);

  const toggleModal = () => {
    dispatch(setIsShowing(!isShowing));
  };

  return {
    isShowing,
    toggleModal,
  };
};
