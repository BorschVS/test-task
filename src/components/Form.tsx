import { useDispatch } from 'react-redux';
import { useFormik } from 'formik';
import * as Yup from 'yup';

import { Button, TextField } from '@mui/material';

import { useModal } from 'hooks/useModal';

import { setFormData } from '../redux/ducks/form';
import { AppDispatch } from 'redux/configureStore';

import { FormStyled } from 'styled/Form.styled';

const Form = () => {
  const dispatch = useDispatch<AppDispatch>();

  const { toggleModal } = useModal();

  const formik = useFormik({
    initialValues: {
      name: '',
      phone: '',
      email: '',
    },
    validationSchema: Yup.object({
      name: Yup.string().required('Введите имя'),
      phone: Yup.number().required('Введите номер'),
      email: Yup.string()
        .email('Введите корректный email')
        .required('Введите email'),
    }),
    onSubmit: (values) => {
      dispatch(setFormData(JSON.stringify(values)));
      toggleModal();
    },
  });

  return (
    <FormStyled onSubmit={formik.handleSubmit}>
      <TextField
        fullWidth
        margin="dense"
        id="name"
        name="name"
        label="Имя и фамилия"
        value={formik.values.name}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.name && Boolean(formik.errors.name)}
        helperText={formik.touched.name && formik.errors.name}
      />
      <TextField
        fullWidth
        margin="dense"
        id="phone"
        name="phone"
        label="Телефон"
        value={formik.values.phone}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.phone && Boolean(formik.errors.phone)}
        helperText={formik.errors.phone && 'Введите номер'}
      />
      <TextField
        fullWidth
        margin="dense"
        id="email"
        name="email"
        label="Почта"
        value={formik.values.email}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.email && Boolean(formik.errors.email)}
        helperText={formik.touched.email && formik.errors.email}
      />
      <Button
        sx={{ marginTop: 2 }}
        color="info"
        variant="contained"
        fullWidth
        type="submit"
      >
        Отправить
      </Button>
    </FormStyled>
  );
};

export default Form;
