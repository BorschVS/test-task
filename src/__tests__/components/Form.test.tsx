import '@testing-library/jest-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import { setFormData } from '../../redux/ducks/formSlice';
import Form from 'components/Form';
import { useModal } from 'hooks/useModal';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

jest.mock('../../redux/ducks/formSlice', () => ({
  setFormData: jest.fn(),
}));

jest.mock('hooks/useModal', () => ({
  useModal: jest.fn(() => ({
    toggleModal: jest.fn(),
  })),
}));

describe('Form component', () => {
  const mockDispatch = jest.fn();
  const mockToggleModal = jest.fn();

  beforeEach(() => {
    (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    (useModal as jest.Mock).mockReturnValue({
      isShowing: false,
      toggleModal: mockToggleModal,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Should render form with initial empty name field', () => {
    render(<Form />);

    const nameFields = screen.getAllByLabelText('Имя и фамилия');
    nameFields.forEach((field) => expect(field).toHaveValue(''));
  });

  test('Should render form with initial empty phone field', () => {
    render(<Form />);

    const phoneFields = screen.getAllByLabelText('Телефон');
    phoneFields.forEach((field) => expect(field).toHaveValue(''));
  });

  test('Should render form with initial empty email field', () => {
    render(<Form />);

    const postFields = screen.getAllByLabelText('Почта');
    postFields.forEach((field) => expect(field).toHaveValue(''));
  });

  test('Should display validation error for empty name field', async () => {
    render(<Form />);

    const submitButton = screen.getByText('Отправить');
    fireEvent.click(submitButton);

    const validationMessage = await screen.findByText('Введите имя');
    expect(validationMessage).toBeInTheDocument();
  });

  test('Should display validation error for empty phone field', async () => {
    render(<Form />);
    const submitButton = screen.getByText('Отправить');
    fireEvent.click(submitButton);

    const validationMessage = await screen.findByText('Введите номер');
    expect(validationMessage).toBeInTheDocument();
  });

  test('Should display validation error for empty email field', async () => {
    render(<Form />);

    const submitButton = screen.getByText('Отправить');
    fireEvent.click(submitButton);

    const validationMessage = await screen.findByText('Введите email');
    expect(validationMessage).toBeInTheDocument();
  });

  test('Should call dispatch on valid submit', async () => {
    render(<Form />);

    const nameField = screen.getByLabelText('Имя и фамилия');
    const phoneField = screen.getByLabelText('Телефон');
    const emailField = screen.getByLabelText('Почта');

    fireEvent.change(nameField, { target: { value: 'John Doe' } });
    fireEvent.change(phoneField, { target: { value: '1234567890' } });
    fireEvent.change(emailField, { target: { value: 'john@example.com' } });

    const submitButton = screen.getByText('Отправить');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledWith(
        setFormData({
          name: 'John Doe',
          phone: '1234567890',
          email: 'john@example.com',
        })
      );
    });
  });

  test('Should call toggleModal on submit', async () => {
    render(<Form />);

    const nameField = screen.getByLabelText('Имя и фамилия');
    const phoneField = screen.getByLabelText('Телефон');
    const emailField = screen.getByLabelText('Почта');

    fireEvent.change(nameField, { target: { value: 'John Doe' } });
    fireEvent.change(phoneField, { target: { value: '1234567890' } });
    fireEvent.change(emailField, { target: { value: 'john@example.com' } });

    const submitButton = screen.getByText('Отправить');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(mockToggleModal).toHaveBeenCalled();
    });
  });

  test('Should match FlightCard snapshot', () => {
    const { asFragment } = render(<Form />);

    expect(asFragment()).toMatchSnapshot();
  });
});
