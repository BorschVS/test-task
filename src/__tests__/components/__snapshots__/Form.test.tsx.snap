// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component Should match FlightCard snapshot 1`] = `
<DocumentFragment>
  <form
    class="sc-braxZu jclqgR"
    data-testid="form"
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="name"
        id="name-label"
      >
        Имя и фамилия
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="name"
          name="name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Имя и фамилия
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="phone"
        id="phone-label"
      >
        Телефон
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="phone"
          name="phone"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Телефон
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="email"
        id="email-label"
      >
        Почта
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="email"
          name="email"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Почта
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo MuiButton-fullWidth css-tet4c5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Отправить
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
